fuel-vm/src/checked_transaction.rs

```diff
@@ -0,0 +1,972 @@
+/// of `Checked` outside the `fuel-tx` crate.
+///
+/// The inner data is immutable to prevent modification to invalidate the checking.
+///
+/// If you need to modify an inner state, you need to get inner values
+/// (via the `Into<(Tx, Tx ::Metadata)>` trait), modify them and check again.
+///
+/// # Dev note: Avoid serde serialization of this type.
+///
+/// Since checked tx would need to be re-validated on deserialization anyways,
+/// it's cleaner to redo the tx check.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct Checked<Tx: IntoChecked> {
+    transaction: Tx,
+    metadata: Tx::Metadata,
+    checks_bitmask: Checks,
+    /// If predicates have been checked, this is how much gas checking them used.
+    /// This must be zero if the predicates have not been checked yet.
+    /// Note that the checkedness of predicates is marked in checks_bitmask.
+    gas_used_by_predicates: Word,
+}
+
+impl<Tx: IntoChecked> Checked<Tx> {
+    pub(crate) fn basic(transaction: Tx, metadata: Tx::Metadata) -> Self {
+        Self {
+            transaction,
+            metadata,
+            checks_bitmask: Checks::Basic,
+            gas_used_by_predicates: 0,
+        }
+    }
+
+    /// Returns reference on inner transaction.
+    pub fn transaction(&self) -> &Tx {
+        &self.transaction
+    }
+
+    /// Returns the metadata generated during the check for transaction.
+    pub fn metadata(&self) -> &Tx::Metadata {
+        &self.metadata
+    }
+
+    /// Returns the bitmask of all passed checks.
+    pub fn checks(&self) -> &Checks {
+        &self.checks_bitmask
+    }
+
+    /// Gas used by predicate checking. If the predicates have not
+    /// been checked, returns zero.
+    pub fn gas_used_by_predicates(&self) -> Word {
+        self.gas_used_by_predicates
+    }
+
+    /// Performs check of signatures, if not yet done.
+    pub fn check_signatures(mut self) -> Result<Self, CheckError> {
+        if !self.checks_bitmask.contains(Checks::Signatures) {
+            self.transaction.check_signatures()?;
+            self.checks_bitmask.insert(Checks::Signatures);
+        }
+        Ok(self)
+    }
+}
+
+// #[cfg(feature = "internals")]
+impl<Tx: IntoChecked + Default> Default for Checked<Tx>
+where
+    Checked<Tx>: CheckPredicates,
+{
+    fn default() -> Self {
+        Tx::default()
+            .into_checked(Default::default(), &Default::default(), Default::default())
+            .expect("default tx should produce a valid fully checked transaction")
+    }
+}
+
+impl<Tx: IntoChecked> From<Checked<Tx>> for (Tx, Tx::Metadata) {
+    fn from(checked: Checked<Tx>) -> Self {
+        let Checked {
+            transaction, metadata, ..
+        } = checked;
+
+        (transaction, metadata)
+    }
+}
+
+impl<Tx: IntoChecked> AsRef<Tx> for Checked<Tx> {
+    fn as_ref(&self) -> &Tx {
+        &self.transaction
+    }
+}
+
+// #[cfg(feature = "internals")]
```

Comment:
```suggestion

```

Reply:
These methods were commented out due to migrating the types from fuel-tx, but they shouldn't be made available as part of the normal public api. I put them under `test-helpers` since we don't want the ability for other crates like fuel-core to have mutable access to checked transactions which could allow for manually flipping validation flags without doing the real validation etc.